### ISSUE TITLE
Nvidia segfault on ShaderMaterial destructor from Mono Finalizer thread.

### ISSUE THREAD

**Issue by PJB3005** [godotengine/godot#24108]
**Godot version:**
5b8dde4b4a5a930e08da51e07ec87f43e5bd7b36

**OS/device including version:**
Arch Linux, Intel UHD 630 with GTX 1050 Ti through Optimus, properietary driver 415.18-4. **This crash only happens on my NVidia card, if I switch back to my iGPU it's fine**.

**Issue description:**
My game crashes when a `ShaderMaterial` is disposed from the Mono finalizer thread. As mention above it only happens on nvidia and on Intel it's fine. Stack trace from lldb:

```
* thread #7, name = 'Finalizer', stop reason = signal SIGSEGV: invalid address (fault address: 0x1060)
  * frame #0: 0x00007ffff5cdb9e9 libGLdispatch.so.0
    frame #1: 0x00000000030983f2 godot.x11.tools.64.mono`VisualServerRaster::free(this=0x0000000004eabe20, p_rid=RID @ 0x00007fffd9ffd830) at visual_server_raster.cpp:72
    frame #2: 0x00000000030e1d42 godot.x11.tools.64.mono`VisualServerWrapMT::free(this=0x0000000004f0b120, p1=RID @ 0x00007fffd9ffd850) at visual_server_wrap_mt.h:579
    frame #3: 0x0000000002ccca4d godot.x11.tools.64.mono`Material::~Material(this=0x0000000009a2a5b0) at material.cpp:101
    frame #4: 0x0000000002ccdf62 godot.x11.tools.64.mono`ShaderMaterial::~ShaderMaterial(this=0x0000000009a2a5b0) at material.cpp:255
    frame #5: 0x00000000011d6574 godot.x11.tools.64.mono`void memdelete<Object>(p_class=0x0000000009a2a5b0) at memory.h:119
    frame #6: 0x00000000016096de godot.x11.tools.64.mono`CSharpInstance::_unreference_owner_unsafe(this=0x0000000009911dc0) at csharp_script.cpp:1466
    frame #7: 0x0000000001609bb3 godot.x11.tools.64.mono`CSharpInstance::mono_object_disposed_baseref(this=0x0000000009911dc0, p_obj=0x00007fffbeb04a90, p_is_finalizer=true, r_owner_deleted=0x00007fffd9ffd967) at csharp_script.cpp:1522
    frame #8: 0x00000000018d2713 godot.x11.tools.64.mono`godot_icall_Reference_Disposed(p_obj=0x00007fffbeb04a90, p_ptr=0x0000000009a2a5b0, p_is_finalizer=true) at base_object_glue.cpp:86
    frame #9: 0x00000000403e3485
    frame #10: 0x00000000403e2ed0
    frame #11: 0x00007ffff6d5f16b libmonosgen-2.0.so.1`___lldb_unnamed_symbol4520$$libmonosgen-2.0.so.1 + 763
    frame #12: 0x00007ffff6d84020 libmonosgen-2.0.so.1`___lldb_unnamed_symbol4969$$libmonosgen-2.0.so.1 + 192
    frame #13: 0x00007ffff6d5f8a5 libmonosgen-2.0.so.1`___lldb_unnamed_symbol4523$$libmonosgen-2.0.so.1 + 485
    frame #14: 0x00007ffff6d0f4bc libmonosgen-2.0.so.1`___lldb_unnamed_symbol3918$$libmonosgen-2.0.so.1 + 300
    frame #15: 0x00007ffff6a92a9d libpthread.so.0`start_thread + 253
    frame #16: 0x00007ffff6692b23 libc.so.6`__GI___clone + 67
```

However.. you'll obviously see that that stack trace is clearly incorrect. It's missing a step, namely `RasterizerStorageGLES3::free` before the entry into the GL library.

I decided to step through with lldb and the crash is happening on line 7377 of `rasterizer_storage_gles3.cpp`:
```
(lldb) n
Process 13354 stopped
* thread #7, name = 'Finalizer', stop reason = step over
    frame #0: 0x0000000001c37cbf godot.x11.tools.64.mono`RasterizerStorageGLES3::free(this=0x0000000004eaed90, p_rid=RID @ 0x00007fffd9ffd710) at rasterizer_storage_gles3.cpp:7377
   7374                 }
   7375
   7376                 if (material->ubo_id) {
-> 7377                         glDeleteBuffers(1, &material->ubo_id);
   7378                 }
   7379
   7380                 //remove from owners
(lldb) n
Process 13354 stopped
* thread #7, name = 'Finalizer', stop reason = signal SIGSEGV: invalid address (fault address: 0x1060)
    frame #0: 0x00007ffff5cdb9e9 libGLdispatch.so.0
->  0x7ffff5cdb9e9: jmpq   *0x1060(%rax)
```

I'm not gonna claim to be experienced with OpenGL, but isn't OpenGL generally extremely non thread safe? This is all happening on the Mono Finalizer thread instead of the main thread. This is reliable by the way, I haven't had it *not* happen yet.

**Comment by cart** [comment]
At this risk of hijacking this issue, I'm getting something very similar (also in the destructor of ShaderMaterial), but I get an unhandled null reference exception on line 57 of Object.base.cs:

```cs
 godot_icall_Reference_Disposed(this, ptr, !disposing);
```

Reverting this commit resolves the problem for me: e558e1ec09aa27852426bbd24dfa21e9b60cfbfc

**Comment by cart** [comment]
Or at least, it used to. Reverting that commit in master is now non trivial.

Here's the branch I'm currently dev-ing on: https://github.com/cart/godot/commits/highhatcustom

**Comment by neikeq** [comment]
This is actually being caused by #23709

**Comment by neikeq** [comment]
Fixed by #25300
